bdk_core/src/sparse_chain.rs

```diff
@@ -211,59 +213,85 @@ impl<I: ChainIndex> SparseChain<I> {
             }
         }
 
+        Ok(invalid_from)
+    }
+
+    pub fn formulate_changeset(&self, update: &Self, invalid_from: u32) -> ChangeSet<I> {
         // create initial change-set, based on checkpoints and txids that are to be invalidated
-        let mut change_set = ChangeSet::<I> {
+        let mut changeset = ChangeSet::<I> {
             checkpoints: self
                 .checkpoints
                 .range(invalid_from..)
                 .map(|(height, _)| (*height, None))
                 .collect(),
             txids: self
-                .ordered_txids
-                // avoid invalidating mempool txids for initial change-set
-                .range(
-                    &(
-                        I::min_ord_of_height(TxHeight::Confirmed(invalid_from)),
-                        Txid::all_zeros(),
-                    )
-                        ..&(
-                            I::min_ord_of_height(TxHeight::Unconfirmed),
-                            Txid::all_zeros(),
-                        ),
-                )
+                .range_txids_by_height(TxHeight::Confirmed(invalid_from)..TxHeight::Unconfirmed)
                 .map(|(_, txid)| (*txid, None))
                 .collect(),
         };
 
         for (&height, &new_hash) in &update.checkpoints {
             let original_hash = self.checkpoints.get(&height).cloned();
 
-            let update_hash = *change_set
+            let update_hash = *changeset
                 .checkpoints
                 .entry(height)
                 .and_modify(|change| *change = Some(new_hash))
                 .or_insert_with(|| Some(new_hash));
 
             if original_hash == update_hash {
-                change_set.checkpoints.remove(&height);
+                changeset.checkpoints.remove(&height);
             }
         }
 
         for (txid, new_index) in &update.txid_to_index {
             let original_index = self.txid_to_index.get(txid).cloned();
 
-            let update_index = change_set
+            let update_index = changeset
                 .txids
                 .entry(*txid)
                 .and_modify(|change| *change = Some(new_index.clone()))
                 .or_insert_with(|| Some(new_index.clone()));
 
             if original_index == *update_index {
-                change_set.txids.remove(txid);
+                changeset.txids.remove(txid);
             }
         }
 
-        Result::Ok(change_set)
+        changeset
+    }
+
+    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet<I>, UpdateFailure<I>> {
+        let invalid_from = self.determine_invalid_from(update)?;
+        let changeset = self.formulate_changeset(update, invalid_from);
+        Result::Ok(changeset)
+    }
+
+    pub fn determine_unconfirming_changeset(
+        &self,
+        graph: &TxGraph,
+        update: &Self,
+        update_graph: &TxGraph,
```

Comment:
What's wrong with passing one graph here?
What's wrong with already applying the update in `update` before doing this step?

I think you are trying to make the update and the unconfirmed consistency stuff all happen in one go. I think this might be a mistake from an API perspective. I think you should be able to take any sparse chain and make the unconfirmed txs consistent against a graph.

Maybe we shouldn't even pass in a graph to do this job. What about something like [`Vec::retain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain). e.g `retain_unconfirmed(&mut self, pred: impl Fn(Txid) -> bool) -> ChangeSet`. Then the caller can use a `TxGraph` or whatever they want to figure out whether something conflicts and whether it should be retained.

Reply:
> I think you should be able to take any sparse chain and make the unconfirmed txs consistent against a graph.

I don't think this is possible, as we will need the total view of the mempool to "evict" the right transactions.

For example, given a confirmed output `A:0`, which is spent by tx `B`. If `B:0` is an unowned output, and there exists a tx `C` which does CPFP from `B:0`, then `C` would not exist in the sparsechain (as it does not contain or reference a relevant scriptPubKey). Now imagine a RBF attempt to replace `B` (called `B'`).Basically, How can we guarantee consistency if we don't have a total view of all descendant transactions?

Reply:
I feel like the best approach is to assume the `update` always contains txs that are either confirmed, or guaranteed to be in the mempool (for that point in time). So anything that conflicts with the update's transactions can be evicted.

Reply:
> Basically, How can we guarantee consistency if we don't have a total view of all descendant transactions?

We're not trying to guarantee it I think but I see what you're getting at. What's the point of doing this if we can't provide guarantees.


> I feel like the best approach is to assume the `update` always contains txs that are either confirmed, or guaranteed to be in the mempool (for that point in time). So anything that conflicts with the update's transactions can be evicted.
> 
> This does not 100% guarantee consistency, but it can guarantee eventual consistency without wrongly evicting transactions.

Great point. Ok I understand your approach much better now.